### ISSUE TITLE
StreamServerInterceptor WithRecoveryHandler not working

### ISSUE THREAD

**Issue by Wander1024** [grpc-ecosystem/go-grpc-middleware#410]
HI. 
I using thanos 0.15.0 version.  it use `go-grpc-middleware v1.2.2` and `go-grpc-prometheus v1.2.0`
it declares the grpc options as following
```
	grpcOpts := []grpc.ServerOption{
		grpc.MaxSendMsgSize(math.MaxInt32),
		grpc_middleware.WithUnaryServerChain(
			met.UnaryServerInterceptor(),
			tracing.UnaryServerInterceptor(tracer),
			grpc_recovery.UnaryServerInterceptor(grpc_recovery.WithRecoveryHandler(grpcPanicRecoveryHandler)),
		),
		grpc_middleware.WithStreamServerChain(
			met.StreamServerInterceptor(),
			tracing.StreamServerInterceptor(tracer),
			grpc_recovery.StreamServerInterceptor(grpc_recovery.WithRecoveryHandler(grpcPanicRecoveryHandler)),
		),
	}
```
but the finishServerSpan func seems panic when doing serverSpan.LogKV(k, vErr.Error())

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x767b2b]

goroutine 48441409129 [running]:

github.com/pkg/errors.(*withStack).Error(0x0, 0x24efb88, 0x5)
        <autogenerated>:1 +0x2b
github.com/grpc-ecosystem/go-grpc-middleware/tracing/opentracing.finishServerSpan(0x2a58980, 0xc41e49a6f0, 0x2a8e960, 0x3f7bc20, 0x29f9d00, 0x0)

        /app/vendor/github.com/grpc-ecosystem/go-grpc-middleware/tracing/opentracing/server_interceptors.go:85 +0x3de
github.com/grpc-ecosystem/go-grpc-middleware/tracing/opentracing.StreamServerInterceptor.func1(0x2362040, 0xc38e362160, 0x2a72200, 0xcb80dd94a0, 0xcb80dd9400, 0xcb80dd9420, 0x2a58980, 0xc41e49a6c0)

        /app/vendor/github.com/grpc-ecosystem/go-grpc-middleware/tracing/opentracing/server_interceptors.go:48 +0x18a
github.com/thanos-io/thanos/pkg/tracing.StreamServerInterceptor.func1(0x2362040, 0xc38e362160, 0x2a72320, 0xcb80dd9480, 0xcb80dd9400, 0xcb80dd9420, 0x29b4a79, 0x0)

        /app/pkg/tracing/grpc.go:42 +0x134
github.com/grpc-ecosystem/go-grpc-middleware.ChainStreamServer.func1.1.1(0x2362040, 0xc38e362160, 0x2a72320, 0xcb80dd9480, 0x14, 0xc994ef1180)

        /app/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:49 +0x5f
github.com/grpc-ecosystem/go-grpc-prometheus.(*ServerMetrics).StreamServerInterceptor.func1(0x2362040, 0xc38e362160, 0x2a72680, 0xc3aba79140, 0xcb80dd9400, 0xcb80dd9440, 0xcb80dd9460, 0xcb80dd9400)

        /app/vendor/github.com/grpc-ecosystem/go-grpc-prometheus/server_metrics.go:121 +0xeb
github.com/grpc-ecosystem/go-grpc-middleware.ChainStreamServer.func1.1.1(0x2362040, 0xc38e362160, 0x2a72680, 0xc3aba79140, 0xca5f92ac68, 0x40cf58)

        /app/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:49 +0x5f
github.com/grpc-ecosystem/go-grpc-middleware.ChainStreamServer.func1(0x2362040, 0xc38e362160, 0x2a72680, 0xc3aba79140, 0xcb80dd9400, 0x2609c68, 0x2a58980, 0xc41e49a660)

        /app/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:58 +0xcf
google.golang.org/grpc.(*Server).processStreamingRPC(0xc0c972b6c0, 0x2a83ba0, 0xd0f7b07500, 0xc3c966aa00, 0xc3821b19b0, 0x3f19120, 0x0, 0x0, 0x0)
        /app/vendor/google.golang.org/grpc/server.go:1336 +0x511

google.golang.org/grpc.(*Server).handleStream(0xc0c972b6c0, 0x2a83ba0, 0xd0f7b07500, 0xc3c966aa00, 0x0)
        /app/vendor/google.golang.org/grpc/server.go:1409 +0xc66

google.golang.org/grpc.(*Server).serveStreams.func1.1(0xd8a9c8b6f0, 0xc0c972b6c0, 0x2a83ba0, 0xd0f7b07500, 0xc3c966aa00)
        /app/vendor/google.golang.org/grpc/server.go:746 +0xa1
        
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /app/vendor/google.golang.org/grpc/server.go:744 +0xa1
```

**Comment by Wander1024** [comment]
sry, this panic caused by
```
	if err != nil {
		ext.Error.Set(serverSpan, true)
		serverSpan.LogFields(log.String("event", "error"), log.String("message", err.Error()))
	}
```

**Comment by Wander1024** [comment]
seemes this panic raised in `tracing.StreamServerInterceptor(tracer)`.
why can't `grpc_recovery.StreamServerInterceptor` catch it?

**Comment by johanbrandhorst** [comment]
Fixed in #414 I assume?

**Comment by Wander1024** [comment]
Yes.
